appscale/tools/appscale_tools.py

```diff
@@ -215,29 +227,193 @@ def add_keypair(cls, options):
 
   @classmethod
   def describe_instances(cls, options):
-    """Queries each node in the currently running AppScale deployment and
-    reports on their status.
+    """
+    Gets cluster stats and prints it nicely.
 
     Args:
       options: A Namespace that has fields for each parameter that can be
         passed in via the command-line interface.
     """
-    login_host = LocalState.get_login_host(options.keyname)
-    login_acc = AppControllerClient(login_host,
-      LocalState.get_secret_key(options.keyname))
+    try:
+      login_host = LocalState.get_login_host(options.keyname)
+      login_acc = AppControllerClient(login_host,
+        LocalState.get_secret_key(options.keyname))
+      all_public_ips = login_acc.get_all_public_ips()
+      cluster_stats = login_acc.get_cluster_stats()
+    except (faultType, AppControllerException, BadConfigurationException):
+      AppScaleLogger.warn("AppScale deployment is probably down")
+      raise
+
+    # Convert cluster stats to useful structures
+    node_stats = {
+      ip: next((n for n in cluster_stats if n["public_ip"] == ip), None)
```

Comment:
It is not the most well readable construction, but I found that many experienced python devs use it and it seems to be like well known way to get the first match in list.
Function `next` returns the next element for iterator. In this case I pass iterator is generator `n for n in cluster_stats if n["public_ip"] == ip`. This generator iterates though list of nodes and yield only those which match public IP.